js_modules/dagster-ui/packages/ui-core/src/instance/VirtualizedInstanceConcurrencyTable.tsx

```diff
@@ -0,0 +1,235 @@
+import {gql, useLazyQuery} from '@apollo/client';
+import {
+  Box,
+  Button,
+  ButtonLink,
+  Colors,
+  Icon,
+  Menu,
+  MenuItem,
+  Popover,
+  Tag,
+} from '@dagster-io/ui-components';
+import {useVirtualizer} from '@tanstack/react-virtual';
+import {useRef} from 'react';
+import styled from 'styled-components';
+
+import {
+  SingleConcurrencyKeyQuery,
+  SingleConcurrencyKeyQueryVariables,
+} from './types/VirtualizedInstanceConcurrencyTable.types';
+import {FIFTEEN_SECONDS, useQueryRefreshAtInterval} from '../app/QueryRefresh';
+import {Container, HeaderCell, Inner, Row, RowCell} from '../ui/VirtualizedTable';
+import {LoadingOrNone, useDelayedRowQuery} from '../workspace/VirtualizedWorkspaceTable';
+
+const TEMPLATE_COLUMNS = '1fr 150px 150px 150px 150px 150px';
+
+export const ConcurrencyTable = ({
+  concurrencyKeys,
+  onEdit,
+  onDelete,
+  onSelect,
+}: {
+  concurrencyKeys: string[];
+  onEdit: (key: string) => void;
+  onDelete: (key: string) => void;
+  onSelect: (key: string | undefined) => void;
+}) => {
+  const parentRef = useRef<HTMLDivElement | null>(null);
+
+  const rowVirtualizer = useVirtualizer({
+    count: concurrencyKeys.length,
+    getScrollElement: () => parentRef.current,
+    estimateSize: () => 64,
+    overscan: 10,
+  });
+
+  const totalHeight = rowVirtualizer.getTotalSize();
+  const items = rowVirtualizer.getVirtualItems();
+
+  return (
+    <>
+      <ConcurrencyHeader />
+      <div style={{overflow: 'hidden'}}>
+        <Container ref={parentRef}>
+          <Inner $totalHeight={totalHeight}>
+            {items.map(({index, key, size, start}) => {
+              const concurrencyKey: string = concurrencyKeys[index]!;
```

Comment:
Probably shouldn't need to cast this as `string`, which the `!` should do implicitly.